// @rcsb/rcsb-molstar/build/src/viewer/ui/import.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ImportControls = void 0;
var tslib_1 = require("tslib");
var jsx_runtime_1 = require("react/jsx-runtime");
/**
 * Copyright (c) 2019-2020 mol* contributors, licensed under MIT, See LICENSE file for more info.
 *
 * @author Alexander Rose <alexander.rose@weirdbyte.de>
 */
var mol_state_1 = require("molstar/lib/mol-state");
var file_1 = require("molstar/lib/mol-plugin-state/actions/file");
var structure_1 = require("molstar/lib/mol-plugin-state/actions/structure");
var apply_action_1 = require("molstar/lib/mol-plugin-ui/state/apply-action");
var base_1 = require("molstar/lib/mol-plugin-ui/base");
var icons_1 = require("molstar/lib/mol-plugin-ui/controls/icons");
var ImportControls = /** @class */ (function (_super) {
    tslib_1.__extends(ImportControls, _super);
    function ImportControls() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    ImportControls.prototype.defaultState = function () {
        return {
            header: 'Import',
            isCollapsed: false,
            brand: { accent: 'gray', svg: icons_1.FileOutlineSvg }
        };
    };
    ImportControls.prototype.renderControls = function () {
        return (0, jsx_runtime_1.jsxs)("div", tslib_1.__assign({ className: 'msp-control-offset', style: { paddingTop: '1px' } }, { children: [(0, jsx_runtime_1.jsx)(apply_action_1.ApplyActionControl, { state: this.plugin.state.data, action: file_1.OpenFiles, nodeRef: mol_state_1.StateTransform.RootRef, initiallyCollapsed: true }, "".concat(file_1.OpenFiles.id)), (0, jsx_runtime_1.jsx)(apply_action_1.ApplyActionControl, { state: this.plugin.state.data, action: structure_1.DownloadStructure, nodeRef: mol_state_1.StateTransform.RootRef, initiallyCollapsed: true }, "".concat(structure_1.DownloadStructure.id))] }));
    };
    return ImportControls;
}(base_1.CollapsableControls));
exports.ImportControls = ImportControls;
