// @rcsb/rcsb-molstar/build/src/viewer/ui/session.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SessionControls = void 0;
var tslib_1 = require("tslib");
var jsx_runtime_1 = require("react/jsx-runtime");
/**
 * Copyright (c) 2020 mol* contributors, licensed under MIT, See LICENSE file for more info.
 *
 * @author Alexander Rose <alexander.rose@weirdbyte.de>
 */
var base_1 = require("molstar/lib/mol-plugin-ui/base");
var snapshots_1 = require("molstar/lib/mol-plugin-ui/state/snapshots");
var icons_1 = require("molstar/lib/mol-plugin-ui/controls/icons");
var LocalStateControls = /** @class */ (function (_super) {
    tslib_1.__extends(LocalStateControls, _super);
    function LocalStateControls() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    LocalStateControls.prototype.defaultState = function () {
        return {
            header: 'Views',
            isCollapsed: true,
        };
    };
    LocalStateControls.prototype.renderControls = function () {
        return (0, jsx_runtime_1.jsxs)("div", { children: [(0, jsx_runtime_1.jsx)(snapshots_1.LocalStateSnapshots, {}), (0, jsx_runtime_1.jsx)(snapshots_1.LocalStateSnapshotList, {})] });
    };
    return LocalStateControls;
}(base_1.CollapsableControls));
var StateControls = /** @class */ (function (_super) {
    tslib_1.__extends(StateControls, _super);
    function StateControls() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    StateControls.prototype.defaultState = function () {
        return {
            header: 'Download / Open',
            isCollapsed: true,
        };
    };
    StateControls.prototype.renderControls = function () {
        return (0, jsx_runtime_1.jsx)("div", { children: (0, jsx_runtime_1.jsx)(snapshots_1.StateExportImportControls, {}) });
    };
    return StateControls;
}(base_1.CollapsableControls));
var SessionControls = /** @class */ (function (_super) {
    tslib_1.__extends(SessionControls, _super);
    function SessionControls() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    SessionControls.prototype.defaultState = function () {
        return {
            header: 'Session',
            isCollapsed: true,
            brand: { accent: 'gray', svg: icons_1.SaveOutlinedSvg }
        };
    };
    SessionControls.prototype.renderControls = function () {
        return (0, jsx_runtime_1.jsxs)("div", tslib_1.__assign({ className: 'msp-control-offset', style: { paddingTop: '1px' } }, { children: [(0, jsx_runtime_1.jsx)(LocalStateControls, {}), (0, jsx_runtime_1.jsx)(StateControls, {})] }));
    };
    return SessionControls;
}(base_1.CollapsableControls));
exports.SessionControls = SessionControls;
